daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs

```diff
@@ -95,20 +95,53 @@ navigatorPortNavigatorArgs (NavigatorPort p) = ["--port", show p]
 navigatorURL :: NavigatorPort -> String
 navigatorURL (NavigatorPort p) = "http://localhost:" <> show p
 
-withSandbox :: SandboxClassic -> Maybe SandboxPortSpec -> [String] -> (Process () () () -> SandboxPort -> IO a) -> IO a
-withSandbox (SandboxClassic classic) mbPortSpec extraArgs a = withTempDir $ \tempDir -> do
-    let portFile = tempDir </> "sandbox-portfile"
-    let sandbox = if classic then "sandbox-classic" else "sandbox"
-    let args = concat
-          [ [ sandbox ]
-          , concat [ [ "--port", show (fromSandboxPortSpec portSpec) ] | Just portSpec <- [mbPortSpec] ]
-          , [ "--port-file", portFile ]
-          , extraArgs
-          ]
-    withPlatformJar args "sandbox-logback.xml" $ \ph -> do
-        putStrLn "Waiting for sandbox to start: "
-        port <- readPortFile maxRetries portFile
-        a ph (SandboxPort port)
+-- | Use SandboxPortSpec to determine a sandbox port number.
+-- This is racy thanks to getFreePort, but there's no good alternative at the moment.
+getPortForSandbox :: Int -> Maybe SandboxPortSpec -> IO Int
+getPortForSandbox defaultPort = \case
+    Nothing -> pure defaultPort
+    Just (SpecifiedPort port) -> pure (unSandboxPort port)
+    Just FreePort -> fromIntegral <$> getFreePort
+
+determineCantonPortsLabouriously :: StartOptions -> IO CantonPorts
+determineCantonPortsLabouriously StartOptions{..} = do
+    ledgerApi <- getPortForSandbox 6865 sandboxPortM
+    adminApi <- getPortForSandbox 6866 cantonAdminApiPort
+    domainPublicApi <- getPortForSandbox 6867 cantonDomainPublicPort
+    domainAdminApi <- getPortForSandbox 6868 cantonDomainAdminPort
+    pure CantonPorts {..}
+
+withSandbox :: StartOptions -> FilePath -> [String] -> [String] -> (Process () () () -> SandboxPort -> IO a) -> IO a
+withSandbox startOptions@StartOptions{..} darPath scenarioArgs sandboxArgs kont =
+    case sandboxChoice of
+      SandboxClassic -> oldSandbox "sandbox-classic"
+      SandboxModern -> oldSandbox "sandbox"
+      SandboxCanton -> cantonSandbox
+
+  where
+    cantonSandbox = do
+      cantonPorts <- determineCantonPortsLabouriously startOptions
+      withCantonSandbox cantonPorts sandboxArgs $ \ph -> do
+        let sandboxPort = ledgerApi cantonPorts
+        putStrLn "Waiting for canton sandbox to start: "
+        waitForConnectionOnPort (putStr "." *> threadDelay 500000) sandboxPort
```

Comment:
Not part of this PR but this is unreliable afaict because it can happen before the participant has connected to the domain. I suggest asking the Canton team what they recommend here.